gensim/models/poincare.py

```diff
@@ -0,0 +1,637 @@
+            Number of epochs to use for burn-in initialization (0 means no burn-in), defaults to 0.
+        burn_in_alpha : float, optional
+            learning rate for burn-in initialization, defaults to 0.01, ignored if `burn_in` is 0.
+        encoding : str, optional
+            encoding of training file, defaults to utf8.
+        seed : int, optional
+            seed for random to ensure reproducibility, defaults to 0.
+
+        Returns
+        --------
+        PoincareModel instance
+
+        """
+        self.train_file = train_file
+        self.encoding = encoding
+        self.wv = KeyedVectors()
+        self.size = size
+        self.train_alpha = alpha  # Learning rate for training
+        self.burn_in_alpha = burn_in_alpha  # Learning rate for burn-in
+        self.alpha = alpha  # Current learning rate
+        self.negative = negative
+        self.iter = iter
+        self.workers = workers
+        self.epsilon = epsilon
+        self.burn_in = burn_in
+        self.seed = seed
+        self.random = random.Random(seed)
+        self.np_random = np_random.RandomState(seed)
+        self.init_range = (-0.001, 0.001)
+        self.loss_grad = grad(PoincareModel.loss_fn)
+        self.load_relations()
+        self.init_embeddings()
+
+    def load_relations(self):
+        """Load relations from the train file and build vocab."""
+        vocab = {}
+        index2word = []
+        all_relations = []  # List of all relation pairs
+        term_relations = defaultdict(set)  # Mapping from node index to its related node indices
+
+        with smart_open(self.train_file, 'r', encoding=self.encoding) as f:
+            reader = csv.reader(f, delimiter='\t')
+            for row in reader:
+                assert len(row) == 2, 'Relation pair has more than two items'
+                for item in row:
+                    if item in vocab:
+                        vocab[item].count += 1
+                    else:
+                        vocab[item] = Vocab(count=1, index=len(index2word))
+                        index2word.append(item)
+                node_1, node_2 = row
+                node_1_index, node_2_index = vocab[node_1].index, vocab[node_2].index
+                term_relations[node_1_index].add(node_2_index)
+                relation = (node_1_index, node_2_index)
+                all_relations.append(relation)
+
+        self.wv.vocab = vocab
+        self.wv.index2word = index2word
+        self.indices_set = set((range(len(index2word))))  # Set of all node indices
+        self.indices_array = np.array(range(len(index2word)))  # Numpy array of all node indices
+        counts = np.array([self.wv.vocab[index2word[i]].count for i in range(len(index2word))])
+        self.node_probabilities = counts / counts.sum()
+        self.all_relations = all_relations
+        self.term_relations = term_relations
+        self.negatives_buffer = []  # Buffer to store negative samples, to reduce calls to sampling method
```

Comment:
It would be nice to encapsulate the buffer to a separate class. Would remove some complexity from this class.

Reply:
It's essentially just a list/numpy array + an index, but yes, it does make it conceptually simpler, and the code more readable. Done.

Reply:
Just confirming if something like this was what you meant - https://github.com/RaRe-Technologies/gensim/pull/1696/files#diff-d698c7f682aaf0f30c4694aa584a8faaR718

Reply:
Yes, that's pretty much what I meant. The name is good. Nice.